### PR TITLE
Make span model serializable for use with infinispan cache

### PR DESCRIPTION
**Pull request by pavolloffay**
https://issues.jboss.org/browse/HWKAPM-588

### CODE REVIEW

**Comment by pavolloffay** [comment]
@objectiser could you please review?

File: server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java [hunk 1]
```diff
@@ -55,4 +65,62 @@ public void testPartialUrl() throws MalformedURLException {
...
+        binaryAnnotation.setValue("GET");
+        binaryAnnotation.setEndpoint(createEndpoint("bonjour", "127.0.0.2", 8090));
+        binaryAnnotation.setType(AnnotationType.STRING);
+
+        Span span = new Span(Arrays.asList(binaryAnnotation));
+        span.setAnnotations(serverAnnotations());
+
+        Span deserializedSPan = (Span)deserialize(serialize(span));
+
+        Assert.assertEquals(span.binaryAnnotationMapping(), deserializedSPan.binaryAnnotationMapping());
```
Comment (objectiser) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: Why have three assertEquals - wouldn't it be better to just have one checking 'span' against 'deserializedSPan'? (possibly with lowercase P :)

File: server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java [hunk 2]
```diff
@@ -55,4 +65,62 @@ public void testPartialUrl() throws MalformedURLException {
...
 
+    @Test
+    public void testSerialization() throws IOException, ClassNotFoundException {
+        BinaryAnnotation binaryAnnotation = new BinaryAnnotation();
+        binaryAnnotation.setKey("http.method");
+        binaryAnnotation.setValue("GET");
+        binaryAnnotation.setEndpoint(createEndpoint("bonjour", "127.0.0.2", 8090));
+        binaryAnnotation.setType(AnnotationType.STRING);
+
+        Span span = new Span(Arrays.asList(binaryAnnotation));
```
Comment (objectiser) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: Would it be possible to have a test mapping config, with some more binary annotations, to ensure the endpoint or component type are filled in, as well as details and properties?

**Comment by objectiser** [comment]
@pavolloffay review finished - only a couple of comments. Will need to wait for merging until after 0.10.0 released.

File: server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java (hunk 1 above)
Comment (pavolloffay) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: I didn't want to add equals to Span class but if you want I can do it.

File: server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java (hunk 2 above)
Comment (pavolloffay) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: NodeDetails are created in this test.

File: server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java (hunk 1 above)
Comment (pavolloffay) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: as we discussed: by default all binary annotations are converted to node details.

File: server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java (hunk 2 above)
Comment (pavolloffay) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: as we discussed: by default all binary annotations are converted to node details.

Comment (pavolloffay) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: This test should test if the class with its members is serializable. If node details were serialized then all members of Mapping results are serialized.

**Comment by pavolloffay** [comment]
@objectiser PR updated.

File: server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java [hunk 3]
```diff
@@ -55,4 +67,73 @@ public void testPartialUrl() throws MalformedURLException {
...
+        numberAnnotation.setValue("265.66");
+        numberAnnotation.setEndpoint(createEndpoint("bonjour2", "127.0.1.2", 8090));
+        numberAnnotation.setType(AnnotationType.DOUBLE);
+
+        Span span = new Span(Arrays.asList(stringAnnotation, numberAnnotation));
+        span.setAnnotations(serverAnnotations());
+
+        Span deserializedSpan = (Span)deserialize(serialize(span));
+
+        Assert.assertEquals(span.binaryAnnotationMapping(), deserializedSpan.binaryAnnotationMapping());
```
Comment (objectiser) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: The assertion should be checking equivalent of 'span' against 'deserializedSpan' - to make sure both objects are the same - not just some of the fields.

File: server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java [hunk 4]
```diff
@@ -55,4 +67,73 @@ public void testPartialUrl() throws MalformedURLException {
...
+        Assert.assertEquals(new HashSet<>(Arrays.asList(
+                    new AbstractMap.SimpleEntry<>("http.method", "GET"),
+                    new AbstractMap.SimpleEntry<>("foo", "265.66"))),
+                deserializedSpan.binaryAnnotationMapping().getNodeDetails().entrySet());
```
Comment (objectiser) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: Is it possible to have a test mapping to also check for Property objects?

File: server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java (hunk 2 above)
Comment (objectiser) [inline server/api/src/test/java/org/hawkular/apm/server/api/model/zipkin/SpanTest.java]: The problem is that Property is serializable - but if at some point it was accidentally changed (not likely) then this test would not detect it currently.

**Comment by objectiser** [comment]
@pavolloffay LGTM.

**Pull request merged**